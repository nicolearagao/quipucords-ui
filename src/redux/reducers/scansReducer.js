import { reportsTypes, scansTypes } from '../constants';
import { reduxHelpers } from '../common';
import { helpers } from '../../common';

const initialState = {
  mergeDialog: {
    show: false,
    scans: [],
    details: false
  },
  connection: {},
  inspection: {},
  job: {},
  jobs: {},
  action: {},
  selected: {},
  deleted: {},
  expanded: {},
  update: 0,
  view: {}
};

const scansReducer = (state = initialState, action) => {
  switch (action.type) {
    case scansTypes.UPDATE_SCANS:
      return reduxHelpers.setStateProp(
        null,
        {
          update: helpers.getCurrentDate().getTime()
        },
        {
          state,
          reset: false
        }
      );

    case scansTypes.RESET_ACTIONS:
      return reduxHelpers.setStateProp(
        null,
        {
          deleted: {}
        },
        {
          state,
          reset: false
        }
      );

    case scansTypes.MERGE_SCAN_DIALOG_SHOW:
      return reduxHelpers.setStateProp(
        'mergeDialog',
        {
          show: true,
          scans: action.scans,
          details: action.details
        },
        {
          state,
          initialState
        }
      );

    case scansTypes.MERGE_SCAN_DIALOG_HIDE:
      return reduxHelpers.setStateProp(
        'mergeDialog',
        {
          show: false
        },
        {
          state,
          initialState
        }
      );

    case scansTypes.SELECT_SCAN:
      return reduxHelpers.setStateProp(
        'selected',
        {
          [action.item?.id]: action.item
        },
        {
          state,
          reset: false
        }
      );
    case scansTypes.DESELECT_SCAN:
      const itemsToDeselect = {};
      const deselectItems = (Array.isArray(action.item) && action.item) || [action?.item || {}];
      deselectItems.forEach(({ id }) => {
        itemsToDeselect[id] = null;
      });

      return reduxHelpers.setStateProp(
        'selected',
        {
          ...itemsToDeselect
        },
        {
          state,
          reset: false
        }
      );
    case scansTypes.EXPANDED_SCAN:
      return reduxHelpers.setStateProp(
        'expanded',
        {
          [action.item?.id]: action.cellIndex
        },
        {
          state,
          reset: false
        }
      );
    case scansTypes.NOT_EXPANDED_SCAN:
      return reduxHelpers.setStateProp(
        'expanded',
        {
          [action.item?.id]: null
        },
        {
          state,
          reset: false
        }
      );

    default:
      return reduxHelpers.generatedPromiseActionReducer(
        [
          { ref: 'connection', type: scansTypes.GET_SCAN_CONNECTION_RESULTS },
          { ref: 'inspection', type: scansTypes.GET_SCAN_INSPECTION_RESULTS },
          { ref: 'job', type: scansTypes.GET_SCAN_JOB },
          { ref: 'jobs', type: scansTypes.GET_SCAN_JOBS },
          {
            ref: 'action',
            type: [
              reportsTypes.GET_REPORTS_DOWNLOAD,
              scansTypes.CANCEL_SCAN,
              scansTypes.PAUSE_SCAN,
              scansTypes.RESTART_SCAN,
              scansTypes.START_SCAN
            ]
          },
          { ref: 'deleted', type: scansTypes.DELETE_SCAN },
          { ref: 'view', type: scansTypes.GET_SCANS }
        ],
        state,
        action
      );
  }
};

scansReducer.initialState = initialState;

export { scansReducer as default, initialState, scansReducer };
