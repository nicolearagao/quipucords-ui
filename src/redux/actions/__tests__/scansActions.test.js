import promiseMiddleware from 'redux-promise-middleware';
import { applyMiddleware, combineReducers, legacy_createStore as createStore } from 'redux';
import moxios from 'moxios';
import { multiActionMiddleware } from '../../middleware';
import { scansReducer, scansEditReducer, sourcesReducer } from '../../reducers';
import { scansActions } from '..';

describe('ScansActions', () => {
  const middleware = [multiActionMiddleware, promiseMiddleware];
  const generateStore = () =>
    createStore(
      combineReducers({
        scans: scansReducer,
        scansEdit: scansEditReducer,
        sources: sourcesReducer
      }),
      applyMiddleware(...middleware)
    );

  const generateDispatch = async dispatcher => {
    const store = generateStore();
    return dispatcher(store.dispatch).then(_ => ({ state: store.getState(), value: _ }));
  };

  beforeEach(() => {
    moxios.install();

    moxios.stubRequest(/\/(api).*?/, {
      status: 200,
      timeout: 1,
      response: {
        test: 'success'
      }
    });
  });

  afterEach(() => {
    moxios.uninstall();
  });

  it('Should return response content for addScan method', async () => {
    const dispatcher = scansActions.addScan();

    const { state } = await generateDispatch(dispatcher);
    expect(state.scansEdit.add).toEqual(true);
  });

  it('Should return response content for addStartScan method', async () => {
    const dispatcher = scansActions.addStartScan();

    const { value } = await generateDispatch(dispatcher);
    expect(value.action.type).toMatchSnapshot('addStartScan');
  });

  it('Should return response content for getScans method', async () => {
    const dispatcher = scansActions.getScans();

    const { state } = await generateDispatch(dispatcher);
    expect(state.sources.view.fulfilled).toEqual(true);
    expect(state.scans.view.fulfilled).toEqual(true);
  });

  it('Should return response content for getScanJobs method', async () => {
    const dispatcher = scansActions.getScanJobs('lorem');

    const { state } = await generateDispatch(dispatcher);
    expect(state.scans.jobs.lorem.fulfilled).toEqual(true);
  });

  it('Should return response content for getScanJob method', async () => {
    const dispatcher = scansActions.getScanJob('lorem');

    const { state } = await generateDispatch(dispatcher);
    expect(state.scans.job.lorem.fulfilled).toEqual(true);
  });

  it('Should return response content for getConnectionScanResults method', async () => {
    const dispatcher = scansActions.getConnectionScanResults('lorem');

    const { state } = await generateDispatch(dispatcher);
    expect(state.scans.connection.lorem.fulfilled).toEqual(true);
  });

  it('Should return response content for getInspectionScanResults method', async () => {
    const dispatcher = scansActions.getInspectionScanResults('lorem');

    const { state } = await generateDispatch(dispatcher);
    expect(state.scans.inspection.lorem.fulfilled).toEqual(true);
  });

  it('Should return response content for startScan method', async () => {
    const dispatcher = scansActions.startScan('lorem');

    const { state } = await generateDispatch(dispatcher);
    expect(state.scans.action.lorem.fulfilled).toEqual(true);
  });

  it('Should return response content for pauseScan method', async () => {
    const dispatcher = scansActions.pauseScan('lorem');

    const { state } = await generateDispatch(dispatcher);
    expect(state.scans.action.lorem.fulfilled).toEqual(true);
  });

  it('Should return response content for cancelScan method', async () => {
    const dispatcher = scansActions.cancelScan('lorem');

    const { state } = await generateDispatch(dispatcher);
    expect(state.scans.action.lorem.fulfilled).toEqual(true);
  });

  it('Should return response content for restartScan method', async () => {
    const dispatcher = scansActions.restartScan('lorem');

    const { state } = await generateDispatch(dispatcher);
    expect(state.scans.action.lorem.fulfilled).toEqual(true);
  });

  it('Should return response content for deleteScan method', async () => {
    const dispatcher = scansActions.deleteScan();
    const { value } = await generateDispatch(dispatcher);
    expect(value.action.type).toMatchSnapshot('deleteScan');
  });
});
