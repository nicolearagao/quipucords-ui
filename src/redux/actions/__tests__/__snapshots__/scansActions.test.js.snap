// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ScansActions Should return response content for addStartScan method: addStartScan 1`] = `"ADD_START_SCAN_FULFILLED"`;

exports[`ScansActions Should return response content for deleteScan method: deleteScan 1`] = `"DELETE_SCAN_FULFILLED"`;
